### PR TITLE
Select Facility Flow

### PR DESCRIPTION
**Pull request by ahujamanish**
**This PR includes following updates**

- Added ability to fetch and paginate list of facilities

**What's missing?**
- Ability to deactivate device; need to be handled in the left view

### CODE REVIEW

File: Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		self.tableDataSource.dataSource.innerDataSource.value = viewModel.dataSource
+		self.viewModel.fetchFacilities()
+	}
+
+	override func didReceiveMemoryWarning() {
+		super.didReceiveMemoryWarning()
+	}
+
+	@IBAction private func logoutTapped() {
+		//TODO: Integrate logout call once user management is completed
```
Comment (tinman-fueled) [inline Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Fueled/avingo-ios/blob/c2468198c2827af176fa690eb1d67537fc2b68fc/Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift#L45"></span>Todo Violation: TODOs should be resolved (Integrate logout call once use…). (todo)</td>
    </tr>
  </tbody>
</table>

**Comment by tinman-fueled** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">c2468198c2827af176fa690eb1d67537fc2b68fc <code>chore(device setup/avic-6): merge conflicts</code><br />
• Scope should be all lowercase separated by dashes (given: <code>device setup</code>, expected: <code>device-setup</code>).<br />
• Jira story should be in all uppercase.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">46bbee2f5e45c7912e510b41f37e59db401121f7 <code>feat(device setup/avic-6): add option to logout</code><br />
• Scope should be all lowercase separated by dashes (given: <code>device setup</code>, expected: <code>device-setup</code>).<br />
• Jira story should be in all uppercase.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">ab1bdccc493761e15555c67a4a9f6ddd6b442e28 <code>feat(device setup/avic-6): display list of available facilities to select once logged in</code><br />
• Scope should be all lowercase separated by dashes (given: <code>device setup</code>, expected: <code>device-setup</code>).<br />
• Jira story should be in all uppercase.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift (hunk 1 above)
Comment (harishsaini) [reply Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift]: @ahujamanish logout call is present in the apiclient, we can call and post a logout notification in the app to clear everything

File: Avingo/Code/Helpers/Extensions/UIView+Avingo.swift [hunk 2]
```diff
@@ -11,9 +11,9 @@ import UIKit
 
 extension UIView {
 	func addDropShadow() {
-		self.layer.shadowColor = UIColor.black.cgColor
-		self.layer.shadowRadius = 1.0
-		self.layer.shadowOpacity = 0.10
+		self.layer.shadowColor = UIColor.Avingo.shadowColor.cgColor
+		self.layer.shadowRadius = 7.0
+		self.layer.shadowOpacity = 0.53
```
Comment (harishsaini) [inline Avingo/Code/Helpers/Extensions/UIView+Avingo.swift]: 👍

File: Avingo/Code/Helpers/Protocols/CellNibLoadable.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		let nib = UINib(nibName: self.nibName, bundle: Bundle.main)
+		collectionView.register(nib, forCellWithReuseIdentifier: self.cellIdentifier)
+	}
+}
+
+extension CellNibLoadable where Self: UITableViewHeaderFooterView {
+	static func register(with tableView: UITableView) {
+		let nib = UINib(nibName: self.nibName, bundle: Bundle.main)
+		tableView.register(nib, forHeaderFooterViewReuseIdentifier: self.cellIdentifier)
+	}
```
Comment (harishsaini) [inline Avingo/Code/Helpers/Protocols/CellNibLoadable.swift]: awesome

File: Avingo/Code/Networking/ApiClient.swift [hunk 4]
```diff
@@ -95,13 +95,18 @@ extension ApiClient {
...
 			method: .post,
 			parameters: params,
 			encoding: JSONEncoding.default,
 			headers: nil
 		))
+
+		return userResponseProducer.map { userResponse -> UserResponse in
+			self.authToken.value = userResponse.authToken
+			return userResponse
+		}
```
Comment (harishsaini) [inline Avingo/Code/Networking/ApiClient.swift]: we should remove this change and verify, everything should be working fine on develop

File: Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+			}
+			return SelectFacilityTableViewCell.cellIdentifier
+		}
+		self.tableDataSource.dataSource.innerDataSource.value = viewModel.dataSource
+		self.viewModel.fetchFacilities()
+	}
+
+	override func didReceiveMemoryWarning() {
+		super.didReceiveMemoryWarning()
+	}
```
Comment (harishsaini) [inline Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift]: empty methods should be removed,

i am not very strict about it but i have been asked to do so, up for you to decide

File: Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+			self.fetchFacilitiesAction.values,
+			self.fetchNextPageOfFacilitiesAction.values)
+
+		self.reactive.makeBindingTarget { (model, facilitiesResponse: FacilitiesResponse) in
+			model.continuationToken = facilitiesResponse.next
+			model.canLoadMore.value = !(model.continuationToken ?? "").isEmpty
+			if let result = facilitiesResponse.results {
+				model.facilities.value.append(contentsOf: result)
+			}
+			} <~ response
```
Comment (harishsaini) [inline Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewModel.swift]: 👍

File: Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+	func fetchFacilities() {
+		self.fetchFacilitiesAction.apply(pageSize).start()
+	}
+
+	func fetchNextPageOfFacilities() {
+		guard let continuationToken = self.continuationToken else {
+			return
+		}
+		self.fetchNextPageOfFacilitiesAction.apply(continuationToken).start()
```
Comment (harishsaini) [inline Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewModel.swift]: Everything looks great to me, If we can add an alert when fetching first page fails and a basic activity loader like we have on login to indicate loading state, that would be great and make the controller complete  @ahujamanish

File: Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift (hunk 1 above)
Comment (ahujamanish) [reply Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift]: Ok, I don't think we need logout on this screen anymore.

File: Avingo/Code/Networking/ApiClient.swift (hunk 4 above)
Comment (ahujamanish) [reply Avingo/Code/Networking/ApiClient.swift]: 👍

File: Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift (hunk 5 above)
Comment (ahujamanish) [reply Avingo/Code/UI/Admin/SelectFacility/SelectFacilityViewController.swift]: yeah, makes sense.

**Comment by ahujamanish** [comment]
@harishsaini done!

**Pull request merged**